### ISSUE TITLE
Is the trick for hard gumbel_softmax is correct?

### ISSUE THREAD

**Issue by root20** [pytorch/pytorch#68941]
## To Reproduce

I think the code in the link is not correct.
https://github.com/pytorch/pytorch/blob/48771d1c7f01e5b26c5bf0d6d8973884f908453f/torch/nn/functional.py#L1844

## Expected behavior

To my understanding, the original implementation will compute gradient of (y_hard+y_soft), but I think it should compute the gradient of y_soft.

So, I think the code needs to be changed as follows
`ret = (y_hard - y_soft).detach() + y_soft`
